File: codigo/Back/schemas/demanda.py
from datetime import date
from typing import Optional

from pydantic import BaseModel, ConfigDict

from models.demanda import DemandaStatus


class DemandaBase(BaseModel):
    titulo: str
    descricao: str
    origem: str
    destino: Optional[str] = None
    area_hectares: float
    valor_recompensa: float
    tipo_servico: str
    data_limite: Optional[date] = None


class DemandaCreate(DemandaBase):
    pass


class DemandaResponse(DemandaBase):
    id: int
    status: DemandaStatus

    model_config = ConfigDict(from_attributes=True)


class DemandaStatusUpdate(BaseModel):
    status: DemandaStatus

File: codigo/Mobile/field_flow/lib/domain/entities/demanda.dart
import 'enums.dart';

/// Solicitacao de servico criada por um cliente.
///
/// Entidade de dominio: pura (sem serializacao). A conversao de/para JSON
/// vive em `data/models/demanda_model.dart`.
class Demanda {
  Demanda({
    required this.id,
    required this.clienteId,
    required this.titulo,
    required this.descricao,
    required this.origem,
    required this.areaHectares,
    required this.unidadePagamento,
    required this.tipoServico,
    required this.status,
    this.destino,
    this.origemLat,
    this.origemLng,
    this.destinoLat,
    this.destinoLng,
    this.valorRecompensa,
    this.dataLimite,
    this.prestadorId,
  });

  final int id;
  final int clienteId;
  final String titulo;
  final String descricao;
  final String origem;
  final double areaHectares;
  final UnidadePagamento unidadePagamento;
  final String tipoServico;
  final DemandaStatus status;
  final String? destino;
  final double? origemLat;
  final double? origemLng;
  final double? destinoLat;
  final double? destinoLng;
  final double? valorRecompensa;
  final DateTime? dataLimite;
  final int? prestadorId;

  /// `true` quando ha coordenadas do local da tarefa (origem) para o mapa.
  bool get temLocalizacao => origemLat != null && origemLng != null;
}
